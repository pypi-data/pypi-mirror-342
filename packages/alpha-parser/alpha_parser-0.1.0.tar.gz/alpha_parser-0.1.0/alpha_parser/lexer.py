# Alias module for lexer
from .alpha_lexer import AlphaLexer as Lexer

__all__ = ["Lexer"] 