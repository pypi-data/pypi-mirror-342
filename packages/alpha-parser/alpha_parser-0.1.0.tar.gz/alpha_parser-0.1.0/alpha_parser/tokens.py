from enum import Enum
from typing import Any

class TokenType(Enum):
    # 기본 토큰
    EOF = 'EOF'
    NUMBER = 'NUMBER'
    IDENTIFIER = 'IDENTIFIER'
    
    # 산술 연산자
    PLUS = 'PLUS'
    MINUS = 'MINUS'
    MULTIPLY = 'MULTIPLY'
    DIVIDE = 'DIVIDE'
    POWER = 'POWER'
    
    # 비교 연산자
    EQUAL = 'EQUAL'
    NOT_EQUAL = 'NOT_EQUAL'
    GREATER = 'GREATER'
    LESS = 'LESS'
    GREATER_EQUAL = 'GREATER_EQUAL'
    LESS_EQUAL = 'LESS_EQUAL'
    
    # 논리 연산자
    AND = '&&'
    OR = '||'
    NOT = 'NOT'
    
    # 기타
    LPAREN = 'LPAREN'
    RPAREN = 'RPAREN'
    COMMA = 'COMMA'
    QUESTION = 'QUESTION'
    COLON = 'COLON'
    LBRACKET = 'LBRACKET'
    RBRACKET = 'RBRACKET'

# 알파 공식에서 자주 사용되는 변수들
KNOWN_VARIABLES = {
    'returns', 'open', 'close', 'high', 'low', 'volume', 'vwap', 'cap',
    'close_today', 'sma5', 'sma10', 'sma20', 'sma60',
    'amount', 'turn', 'factor', 'pb', 'pe', 'ps', 'industry'
}

# 알파 공식에서 사용되는 함수들
KNOWN_FUNCTIONS = {
    'rank', 'delay', 'correlation', 'covariance', 'scale', 'delta',
    'signedpower', 'decay_linear', 'indneutralize',
    'ts_min', 'ts_max', 'ts_argmax', 'ts_argmin', 'ts_rank',
    'min', 'max', 'sum', 'product', 'stddev',
    'log', 'abs', 'sign'
}

class Token:
    def __init__(self, type: TokenType, value: Any = None, line: int = 0, column: int = 0):
        self.type = type
        self.value = value
        self.line = line
        self.column = column
        
    def __repr__(self) -> str:
        return f"Token({self.type}, {self.value}, line={self.line}, column={self.column})"
        
    def __eq__(self, other) -> bool:
        if not isinstance(other, Token):
            return False
        return self.type == other.type and self.value == other.value 