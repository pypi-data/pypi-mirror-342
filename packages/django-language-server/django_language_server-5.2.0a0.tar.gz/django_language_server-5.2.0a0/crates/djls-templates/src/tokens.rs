use serde::Serialize;
use std::ops::{Deref, DerefMut};

#[derive(Clone, Debug, Serialize, PartialEq)]
pub enum TokenType {
    Comment(String, String, Option<String>),
    DjangoBlock(String),
    DjangoVariable(String),
    Eof,
    HtmlTagOpen(String),
    HtmlTagClose(String),
    HtmlTagVoid(String),
    Newline,
    ScriptTagOpen(String),
    ScriptTagClose(String),
    StyleTagOpen(String),
    StyleTagClose(String),
    Text(String),
    Whitespace(usize),
}

impl TokenType {
    pub fn len(&self) -> Option<usize> {
        match self {
            TokenType::DjangoBlock(s)
            | TokenType::DjangoVariable(s)
            | TokenType::HtmlTagOpen(s)
            | TokenType::HtmlTagClose(s)
            | TokenType::HtmlTagVoid(s)
            | TokenType::ScriptTagOpen(s)
            | TokenType::ScriptTagClose(s)
            | TokenType::StyleTagOpen(s)
            | TokenType::StyleTagClose(s)
            | TokenType::Text(s) => Some(s.len()),
            TokenType::Comment(content, _, _) => Some(content.len()),
            TokenType::Whitespace(n) => Some(*n),
            TokenType::Newline => Some(1),
            TokenType::Eof => Some(0),
        }
    }
}

#[derive(Clone, Debug, Serialize, PartialEq)]
pub struct Token {
    token_type: TokenType,
    line: usize,
    start: Option<usize>,
}

impl Token {
    pub fn new(token_type: TokenType, line: usize, start: Option<usize>) -> Self {
        Self {
            token_type,
            line,
            start,
        }
    }

    pub fn lexeme(&self) -> String {
        match &self.token_type {
            TokenType::Comment(_, start, end) => match end {
                Some(end) => format!("{} {} {}", start, self.content(), end),
                None => format!("{} {}", start, self.content()),
            },
            TokenType::DjangoBlock(_) => format!("{{% {} %}}", self.content()),
            TokenType::DjangoVariable(_) => format!("{{{{ {} }}}}", self.content()),
            TokenType::Eof => String::new(),
            TokenType::HtmlTagOpen(_)
            | TokenType::ScriptTagOpen(_)
            | TokenType::StyleTagOpen(_) => format!("<{}>", self.content()),
            TokenType::HtmlTagClose(_)
            | TokenType::StyleTagClose(_)
            | TokenType::ScriptTagClose(_) => format!("</{}>", self.content()),
            TokenType::HtmlTagVoid(_) => format!("<{}/>", self.content()),
            TokenType::Newline | TokenType::Text(_) | TokenType::Whitespace(_) => self.content(),
        }
    }

    pub fn content(&self) -> String {
        match &self.token_type {
            TokenType::Comment(s, _, _)
            | TokenType::DjangoBlock(s)
            | TokenType::DjangoVariable(s)
            | TokenType::Text(s)
            | TokenType::HtmlTagOpen(s)
            | TokenType::HtmlTagClose(s)
            | TokenType::HtmlTagVoid(s)
            | TokenType::ScriptTagOpen(s)
            | TokenType::ScriptTagClose(s)
            | TokenType::StyleTagOpen(s)
            | TokenType::StyleTagClose(s) => s.to_string(),
            TokenType::Whitespace(len) => " ".repeat(*len),
            TokenType::Newline => "\n".to_string(),
            TokenType::Eof => "".to_string(),
        }
    }

    pub fn token_type(&self) -> &TokenType {
        &self.token_type
    }

    pub fn line(&self) -> &usize {
        &self.line
    }

    pub fn start(&self) -> Option<u32> {
        self.start.map(|s| s as u32)
    }

    pub fn length(&self) -> Option<u32> {
        self.token_type.len().map(|l| l as u32)
    }

    pub fn is_token_type(&self, token_type: &TokenType) -> bool {
        &self.token_type == token_type
    }
}

#[derive(Clone, Debug, Default, Serialize)]
pub struct TokenStream(Vec<Token>);

impl TokenStream {
    pub fn tokens(&self) -> &Vec<Token> {
        &self.0
    }

    pub fn add_token(&mut self, token: Token) {
        self.0.push(token);
    }

    pub fn finalize(&mut self, line: usize) -> TokenStream {
        let eof_token = Token {
            token_type: TokenType::Eof,
            line,
            start: None,
        };
        self.add_token(eof_token);
        self.clone()
    }
}

impl AsRef<[Token]> for TokenStream {
    fn as_ref(&self) -> &[Token] {
        &self.0
    }
}

impl Deref for TokenStream {
    type Target = Vec<Token>;

    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

impl DerefMut for TokenStream {
    fn deref_mut(&mut self) -> &mut Self::Target {
        &mut self.0
    }
}

impl IntoIterator for TokenStream {
    type Item = Token;
    type IntoIter = std::vec::IntoIter<Self::Item>;

    fn into_iter(self) -> Self::IntoIter {
        self.0.into_iter()
    }
}

impl<'a> IntoIterator for &'a TokenStream {
    type Item = &'a Token;
    type IntoIter = std::slice::Iter<'a, Token>;

    fn into_iter(self) -> Self::IntoIter {
        self.0.iter()
    }
}
