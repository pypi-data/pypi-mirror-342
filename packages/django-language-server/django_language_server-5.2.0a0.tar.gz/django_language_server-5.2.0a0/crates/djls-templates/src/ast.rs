use crate::tokens::Token;
use serde::Serialize;
use thiserror::Error;

#[derive(Clone, Default, Debug, Serialize)]
pub struct NodeList {
    nodes: Vec<Node>,
    line_offsets: LineOffsets,
}

impl NodeList {
    pub fn nodes(&self) -> &Vec<Node> {
        &self.nodes
    }

    pub fn line_offsets(&self) -> &LineOffsets {
        &self.line_offsets
    }

    pub fn add_node(&mut self, node: Node) {
        self.nodes.push(node);
    }

    pub fn set_line_offsets(&mut self, line_offsets: LineOffsets) {
        self.line_offsets = line_offsets
    }

    pub fn finalize(&mut self) -> NodeList {
        self.clone()
    }
}

#[derive(Clone, Default, Debug, Serialize)]
pub struct LineOffsets(pub Vec<u32>);

impl LineOffsets {
    pub fn new() -> Self {
        Self(vec![0])
    }

    pub fn add_line(&mut self, offset: u32) {
        self.0.push(offset);
    }

    pub fn position_to_line_col(&self, position: usize) -> (usize, usize) {
        let position = position as u32;
        let line = match self.0.binary_search(&position) {
            Ok(exact_line) => exact_line,    // Position is at start of this line
            Err(0) => 0,                     // Before first line start
            Err(next_line) => next_line - 1, // We're on the previous line
        };

        // Calculate column as offset from line start
        let col = if line == 0 {
            position as usize
        } else {
            (position - self.0[line]) as usize
        };

        // Convert to 1-based line number
        (line + 1, col)
    }

    pub fn line_col_to_position(&self, line: u32, col: u32) -> u32 {
        // line is 1-based, so subtract 1 to get the index
        self.0[(line - 1) as usize] + col
    }
}

#[derive(Clone, Debug, Serialize)]
pub enum Node {
    Tag {
        name: String,
        bits: Vec<String>,
        span: Span,
    },
    Comment {
        content: String,
        span: Span,
    },
    Text {
        content: String,
        span: Span,
    },
    Variable {
        var: String,
        filters: Vec<String>,
        span: Span,
    },
}

#[derive(Clone, Copy, Debug, Eq, PartialEq, Serialize)]
pub struct Span {
    start: u32,
    length: u32,
}

impl Span {
    pub fn new(start: u32, length: u32) -> Self {
        Self { start, length }
    }

    pub fn start(&self) -> &u32 {
        &self.start
    }

    pub fn length(&self) -> &u32 {
        &self.length
    }
}

impl From<Token> for Span {
    fn from(token: Token) -> Self {
        let start = token.start().unwrap_or(0);
        let length = token.content().len() as u32;
        Span::new(start, length)
    }
}

#[derive(Clone, Debug, Error, Serialize)]
pub enum AstError {
    #[error("Empty AST")]
    EmptyAst,
    #[error("Invalid tag '{tag}' structure: {reason}")]
    InvalidTagStructure {
        tag: String,
        reason: String,
        span: Span,
    },
    #[error("Unbalanced structure: '{opening_tag}' at {opening_span:?} missing closing '{expected_closing}'")]
    UnbalancedStructure {
        opening_tag: String,
        expected_closing: String,
        opening_span: Span,
        closing_span: Option<Span>,
    },
    #[error("Invalid {node_type} node: {reason}")]
    InvalidNode {
        node_type: String,
        reason: String,
        span: Span,
    },
    #[error("Unclosed tag: {0}")]
    UnclosedTag(String),
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::lexer::Lexer;
    use crate::parser::Parser;

    mod line_offsets {
        use super::*;

        #[test]
        fn test_new_starts_at_zero() {
            let offsets = LineOffsets::new();
            assert_eq!(offsets.position_to_line_col(0), (1, 0)); // Line 1, column 0
        }

        #[test]
        fn test_start_of_lines() {
            let mut offsets = LineOffsets::new();
            offsets.add_line(10); // Line 2 starts at offset 10
            offsets.add_line(25); // Line 3 starts at offset 25

            assert_eq!(offsets.position_to_line_col(0), (1, 0)); // Line 1, start
            assert_eq!(offsets.position_to_line_col(10), (2, 0)); // Line 2, start
            assert_eq!(offsets.position_to_line_col(25), (3, 0)); // Line 3, start
        }
    }

    mod spans_and_positions {
        use super::*;

        #[test]
        fn test_variable_spans() {
            let template = "Hello\n{{ user.name }}\nWorld";
            let tokens = Lexer::new(template).tokenize().unwrap();
            let mut parser = Parser::new(tokens);
            let (nodelist, errors) = parser.parse().unwrap();
            assert!(errors.is_empty());

            // Find the variable node
            let nodes = nodelist.nodes();
            let var_node = nodes
                .iter()
                .find(|n| matches!(n, Node::Variable { .. }))
                .unwrap();

            if let Node::Variable { span, .. } = var_node {
                // Variable starts after newline + "{{"
                let (line, col) = nodelist
                    .line_offsets()
                    .position_to_line_col(*span.start() as usize);
                assert_eq!(
                    (line, col),
                    (2, 0),
                    "Variable should start at line 2, col 3"
                );

                assert_eq!(*span.length(), 9, "Variable span should cover 'user.name'");
            }
        }
    }
}
