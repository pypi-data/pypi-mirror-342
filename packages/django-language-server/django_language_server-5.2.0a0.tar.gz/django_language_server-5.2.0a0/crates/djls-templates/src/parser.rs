use crate::ast::{AstError, LineOffsets, Node, NodeList, Span};
use crate::lexer::LexerError;
use crate::tokens::{Token, TokenStream, TokenType};
use thiserror::Error;

pub struct Parser {
    tokens: TokenStream,
    current: usize,
    errors: Vec<ParserError>,
}

impl Parser {
    pub fn new(tokens: TokenStream) -> Self {
        Self {
            tokens,
            current: 0,
            errors: Vec::new(),
        }
    }

    pub fn parse(&mut self) -> Result<(NodeList, Vec<ParserError>), ParserError> {
        let mut nodelist = NodeList::default();
        let mut line_offsets = LineOffsets::new();

        for token in self.tokens.tokens() {
            if let TokenType::Newline = token.token_type() {
                if let Some(start) = token.start() {
                    // Add offset for next line
                    line_offsets.add_line(start + 1);
                }
            }
        }

        self.current = 0;

        while !self.is_at_end() {
            match self.next_node() {
                Ok(node) => {
                    nodelist.add_node(node);
                }
                Err(err) => {
                    if !self.is_at_end() {
                        self.errors.push(err);
                        self.synchronize()?
                    }
                }
            }
        }

        nodelist.set_line_offsets(line_offsets);
        Ok((nodelist.finalize(), std::mem::take(&mut self.errors)))
    }

    fn next_node(&mut self) -> Result<Node, ParserError> {
        let token = self.consume()?;

        match token.token_type() {
            TokenType::Comment(_, open, _) => self.parse_comment(open),
            TokenType::Eof => Err(ParserError::stream_error(StreamError::AtEnd)),
            TokenType::DjangoBlock(_) => self.parse_django_block(),
            TokenType::DjangoVariable(_) => self.parse_django_variable(),
            TokenType::HtmlTagClose(_)
            | TokenType::HtmlTagOpen(_)
            | TokenType::HtmlTagVoid(_)
            | TokenType::Newline
            | TokenType::ScriptTagClose(_)
            | TokenType::ScriptTagOpen(_)
            | TokenType::StyleTagClose(_)
            | TokenType::StyleTagOpen(_)
            | TokenType::Text(_)
            | TokenType::Whitespace(_) => self.parse_text(),
        }
    }

    fn parse_comment(&mut self, open: &str) -> Result<Node, ParserError> {
        // Only treat Django comments as Comment nodes
        if open != "{#" {
            return self.parse_text();
        };

        let token = self.peek_previous()?;

        Ok(Node::Comment {
            content: token.content(),
            span: Span::from(token),
        })
    }

    pub fn parse_django_block(&mut self) -> Result<Node, ParserError> {
        let token = self.peek_previous()?;

        let args: Vec<String> = token
            .content()
            .split_whitespace()
            .map(String::from)
            .collect();
        let name = args.first().ok_or(ParserError::EmptyTag)?.clone();
        let bits = args.into_iter().skip(1).collect();
        let span = Span::from(token);

        Ok(Node::Tag { name, bits, span })
    }

    fn parse_django_variable(&mut self) -> Result<Node, ParserError> {
        let token = self.peek_previous()?;

        let content = token.content();
        let bits: Vec<&str> = content.split('|').collect();
        let var = bits
            .first()
            .ok_or(ParserError::EmptyTag)?
            .trim()
            .to_string();
        let filters = bits
            .into_iter()
            .skip(1)
            .map(|s| s.trim().to_string())
            .collect();
        let span = Span::from(token);

        Ok(Node::Variable { var, filters, span })
    }

    fn parse_text(&mut self) -> Result<Node, ParserError> {
        let token = self.peek_previous()?;

        if token.token_type() == &TokenType::Newline {
            return self.next_node();
        }

        let mut text = token.lexeme();

        while let Ok(token) = self.peek() {
            match token.token_type() {
                TokenType::DjangoBlock(_)
                | TokenType::DjangoVariable(_)
                | TokenType::Comment(_, _, _)
                | TokenType::Newline
                | TokenType::Eof => break,
                _ => {
                    let token_text = token.lexeme();
                    text.push_str(&token_text);
                    self.consume()?;
                }
            }
        }

        let content = match text.trim() {
            "" => return self.next_node(),
            trimmed => trimmed.to_string(),
        };

        let start = token.start().unwrap_or(0);
        let offset = text.find(content.as_str()).unwrap_or(0) as u32;
        let length = content.len() as u32;
        let span = Span::new(start + offset, length);

        Ok(Node::Text { content, span })
    }

    fn peek(&self) -> Result<Token, ParserError> {
        self.peek_at(0)
    }

    fn peek_next(&self) -> Result<Token, ParserError> {
        self.peek_at(1)
    }

    fn peek_previous(&self) -> Result<Token, ParserError> {
        self.peek_at(-1)
    }

    fn peek_at(&self, offset: isize) -> Result<Token, ParserError> {
        let index = self.current as isize + offset;
        self.item_at(index as usize)
    }

    fn item_at(&self, index: usize) -> Result<Token, ParserError> {
        if let Some(token) = self.tokens.get(index) {
            Ok(token.clone())
        } else {
            let error = if self.tokens.is_empty() {
                ParserError::stream_error(StreamError::Empty)
            } else if index < self.current {
                ParserError::stream_error(StreamError::AtBeginning)
            } else if index >= self.tokens.len() {
                ParserError::stream_error(StreamError::AtEnd)
            } else {
                ParserError::stream_error(StreamError::InvalidAccess)
            };
            Err(error)
        }
    }

    fn is_at_end(&self) -> bool {
        self.current + 1 >= self.tokens.len()
    }

    fn consume(&mut self) -> Result<Token, ParserError> {
        if self.is_at_end() {
            return Err(ParserError::stream_error(StreamError::AtEnd));
        }
        self.current += 1;
        self.peek_previous()
    }

    fn backtrack(&mut self, steps: usize) -> Result<Token, ParserError> {
        if self.current < steps {
            return Err(ParserError::stream_error(StreamError::AtBeginning));
        }
        self.current -= steps;
        self.peek_next()
    }

    fn synchronize(&mut self) -> Result<(), ParserError> {
        let sync_types = &[
            TokenType::DjangoBlock(String::new()),
            TokenType::DjangoVariable(String::new()),
            TokenType::Comment(String::new(), String::from("{#"), Some(String::from("#}"))),
            TokenType::Eof,
        ];

        while !self.is_at_end() {
            let current = self.peek()?;
            for sync_type in sync_types {
                if *current.token_type() == *sync_type {
                    return Ok(());
                }
            }
            self.consume()?;
        }
        Ok(())
    }
}

#[derive(Debug)]
pub enum StreamError {
    AtBeginning,
    AtEnd,
    Empty,
    InvalidAccess,
}

#[derive(Debug, Error)]
pub enum ParserError {
    #[error("Unexpected token: expected {expected:?}, found {found} at position {position}")]
    UnexpectedToken {
        expected: Vec<String>,
        found: String,
        position: usize,
    },
    #[error("Invalid syntax: {context}")]
    InvalidSyntax { context: String },
    #[error("Empty tag")]
    EmptyTag,
    #[error("Lexer error: {0}")]
    Lexer(#[from] LexerError),
    #[error("Stream error: {kind:?}")]
    StreamError { kind: StreamError },
    #[error("AST error: {0}")]
    Ast(#[from] AstError),
}

impl ParserError {
    pub fn stream_error(kind: impl Into<StreamError>) -> Self {
        Self::StreamError { kind: kind.into() }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::lexer::Lexer;

    mod html {
        use super::*;
        #[test]
        fn test_parse_html_doctype() {
            let source = "<!DOCTYPE html>";
            let tokens = Lexer::new(source).tokenize().unwrap();
            let mut parser = Parser::new(tokens);
            let (nodelist, errors) = parser.parse().unwrap();
            insta::assert_yaml_snapshot!(nodelist);
            assert!(errors.is_empty());
        }

        #[test]
        fn test_parse_html_tag() {
            let source = "<div class=\"container\">Hello</div>";
            let tokens = Lexer::new(source).tokenize().unwrap();
            let mut parser = Parser::new(tokens);
            let (nodelist, errors) = parser.parse().unwrap();
            insta::assert_yaml_snapshot!(nodelist);
            assert!(errors.is_empty());
        }

        #[test]
        fn test_parse_html_void() {
            let source = "<input type=\"text\" />";
            let tokens = Lexer::new(source).tokenize().unwrap();
            let mut parser = Parser::new(tokens);
            let (nodelist, errors) = parser.parse().unwrap();
            insta::assert_yaml_snapshot!(nodelist);
            assert!(errors.is_empty());
        }
    }

    mod django {
        use super::*;

        #[test]
        fn test_parse_django_variable() {
            let source = "{{ user.name }}";
            let tokens = Lexer::new(source).tokenize().unwrap();
            let mut parser = Parser::new(tokens);
            let (nodelist, errors) = parser.parse().unwrap();
            insta::assert_yaml_snapshot!(nodelist);
            assert!(errors.is_empty());
        }

        #[test]
        fn test_parse_django_variable_with_filter() {
            let source = "{{ user.name|title }}";
            let tokens = Lexer::new(source).tokenize().unwrap();
            let mut parser = Parser::new(tokens);
            let (nodelist, errors) = parser.parse().unwrap();
            insta::assert_yaml_snapshot!(nodelist);
            assert!(errors.is_empty());
        }

        #[test]
        fn test_parse_filter_chains() {
            let source = "{{ value|default:'nothing'|title|upper }}";
            let tokens = Lexer::new(source).tokenize().unwrap();
            let mut parser = Parser::new(tokens);
            let (nodelist, errors) = parser.parse().unwrap();
            insta::assert_yaml_snapshot!(nodelist);
            assert!(errors.is_empty());
        }

        #[test]
        fn test_parse_django_if_block() {
            let source = "{% if user.is_authenticated %}Welcome{% endif %}";
            let tokens = Lexer::new(source).tokenize().unwrap();
            let mut parser = Parser::new(tokens);
            let (nodelist, errors) = parser.parse().unwrap();
            insta::assert_yaml_snapshot!(nodelist);
            assert!(errors.is_empty());
        }

        #[test]
        fn test_parse_django_for_block() {
            let source = "{% for item in items %}{{ item }}{% empty %}No items{% endfor %}";
            let tokens = Lexer::new(source).tokenize().unwrap();
            let mut parser = Parser::new(tokens);
            let (nodelist, errors) = parser.parse().unwrap();
            insta::assert_yaml_snapshot!(nodelist);
            assert!(errors.is_empty());
        }

        #[test]
        fn test_parse_complex_if_elif() {
            let source = "{% if x > 0 %}Positive{% elif x < 0 %}Negative{% else %}Zero{% endif %}";
            let tokens = Lexer::new(source).tokenize().unwrap();
            let mut parser = Parser::new(tokens);
            let (nodelist, errors) = parser.parse().unwrap();
            insta::assert_yaml_snapshot!(nodelist);
            assert!(errors.is_empty());
        }

        #[test]
        fn test_parse_django_tag_assignment() {
            let source = "{% url 'view-name' as view %}";
            let tokens = Lexer::new(source).tokenize().unwrap();
            let mut parser = Parser::new(tokens);
            let (nodelist, errors) = parser.parse().unwrap();
            insta::assert_yaml_snapshot!(nodelist);
            assert!(errors.is_empty());
        }

        #[test]
        fn test_parse_nested_for_if() {
            let source =
                "{% for item in items %}{% if item.active %}{{ item.name }}{% endif %}{% endfor %}";
            let tokens = Lexer::new(source).tokenize().unwrap();
            let mut parser = Parser::new(tokens);
            let (nodelist, errors) = parser.parse().unwrap();
            insta::assert_yaml_snapshot!(nodelist);
            assert!(errors.is_empty());
        }

        #[test]
        fn test_parse_mixed_content() {
            let source = "Welcome, {% if user.is_authenticated %}
    {{ user.name|title|default:'Guest' }}
    {% for group in user.groups %}
        {% if forloop.first %}({% endif %}
        {{ group.name }}
        {% if not forloop.last %}, {% endif %}
        {% if forloop.last %}){% endif %}
    {% empty %}
        (no groups)
    {% endfor %}
{% else %}
    Guest
{% endif %}!";
            let tokens = Lexer::new(source).tokenize().unwrap();
            let mut parser = Parser::new(tokens);
            let (nodelist, errors) = parser.parse().unwrap();
            insta::assert_yaml_snapshot!(nodelist);
            assert!(errors.is_empty());
        }
    }

    mod script {
        use super::*;

        #[test]
        fn test_parse_script() {
            let source = r#"<script type="text/javascript">
    // Single line comment
    const x = 1;
    /* Multi-line
        comment */
    console.log(x);
</script>"#;
            let tokens = Lexer::new(source).tokenize().unwrap();
            let mut parser = Parser::new(tokens);
            let (nodelist, errors) = parser.parse().unwrap();
            insta::assert_yaml_snapshot!(nodelist);
            assert!(errors.is_empty());
        }
    }

    mod style {
        use super::*;

        #[test]
        fn test_parse_style() {
            let source = r#"<style type="text/css">
    /* Header styles */
    .header {
        color: blue;
    }
</style>"#;
            let tokens = Lexer::new(source).tokenize().unwrap();
            let mut parser = Parser::new(tokens);
            let (nodelist, errors) = parser.parse().unwrap();
            insta::assert_yaml_snapshot!(nodelist);
            assert!(errors.is_empty());
        }
    }

    mod comments {
        use super::*;

        #[test]
        fn test_parse_comments() {
            let source = "<!-- HTML comment -->{# Django comment #}";
            let tokens = Lexer::new(source).tokenize().unwrap();
            let mut parser = Parser::new(tokens);
            let (nodelist, errors) = parser.parse().unwrap();
            insta::assert_yaml_snapshot!(nodelist);
            assert!(errors.is_empty());
        }
    }

    mod whitespace {
        use super::*;

        #[test]
        fn test_parse_with_leading_whitespace() {
            let source = "     hello";
            let tokens = Lexer::new(source).tokenize().unwrap();
            let mut parser = Parser::new(tokens);
            let (nodelist, errors) = parser.parse().unwrap();
            insta::assert_yaml_snapshot!(nodelist);
            assert!(errors.is_empty());
        }

        #[test]
        fn test_parse_with_leading_whitespace_newline() {
            let source = "\n     hello";
            let tokens = Lexer::new(source).tokenize().unwrap();
            let mut parser = Parser::new(tokens);
            let (nodelist, errors) = parser.parse().unwrap();
            insta::assert_yaml_snapshot!(nodelist);
            assert!(errors.is_empty());
        }

        #[test]
        fn test_parse_with_trailing_whitespace() {
            let source = "hello     ";
            let tokens = Lexer::new(source).tokenize().unwrap();
            let mut parser = Parser::new(tokens);
            let (nodelist, errors) = parser.parse().unwrap();
            insta::assert_yaml_snapshot!(nodelist);
            assert!(errors.is_empty());
        }

        #[test]
        fn test_parse_with_trailing_whitespace_newline() {
            let source = "hello     \n";
            let tokens = Lexer::new(source).tokenize().unwrap();
            let mut parser = Parser::new(tokens);
            let (nodelist, errors) = parser.parse().unwrap();
            insta::assert_yaml_snapshot!(nodelist);
            eprintln!("{:?}", errors);
            assert!(errors.is_empty());
        }
    }

    mod errors {
        use super::*;

        #[test]
        fn test_parse_unclosed_html_tag() {
            let source = "<div>";
            let tokens = Lexer::new(source).tokenize().unwrap();
            let mut parser = Parser::new(tokens);
            let (nodelist, errors) = parser.parse().unwrap();
            insta::assert_yaml_snapshot!(nodelist);
            assert!(errors.is_empty());
        }

        #[test]
        fn test_parse_unclosed_django_if() {
            let source = "{% if user.is_authenticated %}Welcome";
            let tokens = Lexer::new(source).tokenize().unwrap();
            let mut parser = Parser::new(tokens);
            let (nodelist, errors) = parser.parse().unwrap();
            insta::assert_yaml_snapshot!(nodelist);
            assert!(errors.is_empty()); // Parser doesn't care about semantics at this point
        }

        #[test]
        fn test_parse_unclosed_django_for() {
            let source = "{% for item in items %}{{ item.name }}";
            let tokens = Lexer::new(source).tokenize().unwrap();
            let mut parser = Parser::new(tokens);
            let (nodelist, errors) = parser.parse().unwrap();
            insta::assert_yaml_snapshot!(nodelist);
            assert!(errors.is_empty()); // Parser doesn't care about semantics at this point
        }

        #[test]
        fn test_parse_unclosed_script() {
            let source = "<script>console.log('test');";
            let tokens = Lexer::new(source).tokenize().unwrap();
            let mut parser = Parser::new(tokens);
            let (nodelist, errors) = parser.parse().unwrap();
            insta::assert_yaml_snapshot!(nodelist);
            assert!(errors.is_empty());
        }

        #[test]
        fn test_parse_unclosed_style() {
            let source = "<style>body { color: blue; ";
            let tokens = Lexer::new(source).tokenize().unwrap();
            let mut parser = Parser::new(tokens);
            let (nodelist, errors) = parser.parse().unwrap();
            insta::assert_yaml_snapshot!(nodelist);
            assert!(errors.is_empty());
        }

        #[test]
        fn test_parse_error_recovery() {
            let source = r#"<div class="container">
    <h1>Header</h1>
    {% %}
        {# This if is unclosed which does matter #}
        <p>Welcome {{ user.name }}</p>
        <div>
            {# This div is unclosed which doesn't matter #}
        {% for item in items %}
            <span>{{ item }}</span>
        {% endfor %}
    <footer>Page Footer</footer>
</div>"#;
            let tokens = Lexer::new(source).tokenize().unwrap();
            let mut parser = Parser::new(tokens);
            let (nodelist, errors) = parser.parse().unwrap();
            insta::assert_yaml_snapshot!(nodelist);
            assert_eq!(errors.len(), 1);
            assert!(matches!(&errors[0], ParserError::EmptyTag));
        }
    }

    mod full_templates {
        use super::*;

        #[test]
        fn test_parse_full() {
            let source = r#"<!DOCTYPE html>
<html>
    <head>
        <style type="text/css">
            /* Style header */
            .header { color: blue; }
        </style>
        <script type="text/javascript">
            // Init app
            const app = {
                /* Config */
                debug: true
            };
        </script>
    </head>
    <body>
        <!-- Header section -->
        <div class="header" id="main" data-value="123" disabled>
            {% if user.is_authenticated %}
                {# Welcome message #}
                <h1>Welcome, {{ user.name|title|default:'Guest' }}!</h1>
                {% if user.is_staff %}
                    <span>Admin</span>
                {% else %}
                    <span>User</span>
                {% endif %}
            {% endif %}
        </div>
    </body>
</html>"#;
            let tokens = Lexer::new(source).tokenize().unwrap();
            let mut parser = Parser::new(tokens);
            let (nodelist, errors) = parser.parse().unwrap();
            insta::assert_yaml_snapshot!(nodelist);
            assert!(errors.is_empty());
        }
    }

    mod line_tracking {
        use super::*;

        #[test]
        fn test_parser_tracks_line_offsets() {
            let source = "line1\nline2";
            let tokens = Lexer::new(source).tokenize().unwrap();
            let mut parser = Parser::new(tokens);
            let (nodelist, errors) = parser.parse().unwrap();

            let offsets = nodelist.line_offsets();
            eprintln!("{:?}", offsets);
            assert_eq!(offsets.position_to_line_col(0), (1, 0)); // Start of line 1
            assert_eq!(offsets.position_to_line_col(6), (2, 0)); // Start of line 2
            assert!(errors.is_empty());
        }
    }
}
