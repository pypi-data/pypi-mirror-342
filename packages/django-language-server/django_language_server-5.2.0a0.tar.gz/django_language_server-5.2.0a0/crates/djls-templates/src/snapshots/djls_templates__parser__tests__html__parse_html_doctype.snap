---
source: crates/djls-templates/src/parser.rs
expression: nodelist
---
nodes:
  - Text:
      content: "<!DOCTYPE html>"
      span:
        start: 0
        length: 15
line_offsets:
  - 0
